File: phpbetter/orm/orm3.php
<?php

namespace Framework;

use PDO;
use PDOException;

class ORM
{
    private PDO $connection;
    private string $table;
    private array $conditions = [];
    private array $orderBy = [];
    private int $limit = 0;
    private int $offset = 0;

    public function __construct(string $dsn, string $username, string $password, array $options = [])
    {
        try {
            $this->connection = new PDO($dsn, $username, $password, $options);
            $this->connection->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
        } catch (PDOException $e) {
            throw new RuntimeException("Database connection error: " . $e->getMessage());
        }
    }

    public function table(string $table): self
    {
        $this->table = $table;
        $this->conditions = [];
        $this->orderBy = [];
        $this->limit = 0;
        $this->offset = 0;
        return $this;
    }

    public function where(string $column, string $operator, $value): self
    {
        $this->conditions[] = ["column" => $column, "operator" => $operator, "value" => $value];
        return $this;
    }

    public function orderBy(string $column, string $direction = "ASC"): self
    {
        $this->orderBy[] = "$column $direction";
        return $this;
    }

    public function limit(int $limit): self
    {
        $this->limit = $limit;
        return $this;
    }

    public function offset(int $offset): self
    {
        $this->offset = $offset;
        return $this;
    }

    public function get(): array
    {
        $query = "SELECT * FROM {$this->table}";

        if (!empty($this->conditions)) {
            $query .= " WHERE " . implode(" AND ", array_map(fn($cond) => "{$cond['column']} {$cond['operator']} :{$cond['column']}", $this->conditions));
        }

        if (!empty($this->orderBy)) {
            $query .= " ORDER BY " . implode(", ", $this->orderBy);
        }

        if ($this->limit > 0) {
            $query .= " LIMIT {$this->limit}";
        }

        if ($this->offset > 0) {
            $query .= " OFFSET {$this->offset}";
        }

        $stmt = $this->connection->prepare($query);
        foreach ($this->conditions as $cond) {
            $stmt->bindValue(":" . $cond['column'], $cond['value']);
        }

        $stmt->execute();
        return $stmt->fetchAll(PDO::FETCH_ASSOC);
    }

    public function first(): ?array
    {
        $this->limit(1);
        $result = $this->get();
        return $result[0] ?? null;
    }

    public function insert(string $table, array $data): int
    {
        $columns = implode(", ", array_keys($data));
        $placeholders = implode(", ", array_map(fn($key) => ":$key", array_keys($data)));
        $query = "INSERT INTO $table ($columns) VALUES ($placeholders)";
        $stmt = $this->connection->prepare($query);
        foreach ($data as $key => $value) {
            $stmt->bindValue(":$key", $value);
        }
        $stmt->execute();
        return (int)$this->connection->lastInsertId();
    }

    public function update(string $table, int $id, array $data): bool
    {
        $placeholders = implode(", ", array_map(fn($key) => "$key = :$key", array_keys($data)));
        $query = "UPDATE $table SET $placeholders WHERE id = :id";
        $stmt = $this->connection->prepare($query);
        foreach ($data as $key => $value) {
            $stmt->bindValue(":$key", $value);
        }
        $stmt->bindValue(':id', $id, PDO::PARAM_INT);
        return $stmt->execute();
    }

    public function delete(string $table, int $id): bool
    {
        $stmt = $this->connection->prepare("DELETE FROM $table WHERE id = :id");
        $stmt->bindValue(':id', $id, PDO::PARAM_INT);
        return $stmt->execute();
    }

    public function createTable(string $table, array $columns): bool
    {
        $columnDefinitions = array_map(
            fn($name, $type) => "$name $type",
            array_keys($columns),
            $columns
        );
        $query = "CREATE TABLE IF NOT EXISTS $table (" . implode(", ", $columnDefinitions) . ", id INT AUTO_INCREMENT PRIMARY KEY)";
        return $this->connection->exec($query) !== false;
    }

    public function beginTransaction(): void
    {
        $this->connection->beginTransaction();
    }

    public function commit(): void
    {
        $this->connection->commit();
    }

    public function rollback(): void
    {
        $this->connection->rollBack();
    }
}

abstract class Model
{
    protected ORM $orm;
    protected string $table;

    public function __construct(ORM $orm)
    {
        $this->orm = $orm;
    }

    public function find(int $id): ?array
    {
        return $this->orm->table($this->table)->where('id', '=', $id)->first();
    }

    public function findAll(array $conditions = []): array
    {
        $query = $this->orm->table($this->table);
        foreach ($conditions as $column => $value) {
            $query->where($column, '=', $value);
        }
        return $query->get();
    }

    public function save(array $data): int
    {
        if (isset($data['id'])) {
            $this->orm->update($this->table, $data['id'], $data);
            return $data['id'];
        }
        return $this->orm->insert($this->table, $data);
    }

    public function delete(int $id): bool
    {
        return $this->orm->delete($this->table, $id);
    }

    public static function createTable(ORM $orm, array $columns): bool
    {
        $instance = new static($orm);
        return $orm->createTable($instance->table, $columns);
    }
}

?>

File: pythonbetter/orm3.py
import sqlite3
from typing import Any, Dict, List, Optional

class SQLiteORM:
    def __init__(self, db_path: str):
        self.connection = sqlite3.connect(db_path)
        self.connection.row_factory = sqlite3.Row
        self.cursor = self.connection.cursor()
        self._table = ""
        self._conditions = []
        self._order_by_clauses = []
        self._limit_value = 0
        self._offset_value = 0

    def table(self, table_name: str):
        self._table = table_name
        self._conditions = []
        self._order_by_clauses = []
        self._limit_value = 0
        self._offset_value = 0
        return self

    def where(self, column: str, operator: str, value: Any):
        self._conditions.append((column, operator, value))
        return self

    def order_by(self, column: str, direction: str = "ASC"):
        self._order_by_clauses.append(f"{column} {direction}")
        return self

    def limit(self, limit: int):
        self._limit_value = limit
        return self

    def offset(self, offset: int):
        self._offset_value = offset
        return self

    def get(self) -> List[Dict[str, Any]]:
        query = f"SELECT * FROM {self._table}"

        if self._conditions:
            condition_str = " AND ".join([f"{col} {op} ?" for col, op, _ in self._conditions])
            query += f" WHERE {condition_str}"

        if self._order_by_clauses:
            query += f" ORDER BY {', '.join(self._order_by_clauses)}"

        if self._limit_value:
            query += f" LIMIT {self._limit_value}"

        if self._offset_value:
            query += f" OFFSET {self._offset_value}"

        params = [value for _, _, value in self._conditions]
        self.cursor.execute(query, params)
        return [dict(row) for row in self.cursor.fetchall()]

    def first(self) -> Optional[Dict[str, Any]]:
        self.limit(1)
        results = self.get()
        return results[0] if results else None

    def insert(self, data: Dict[str, Any]) -> int:
        columns = ", ".join(data.keys())
        placeholders = ", ".join(["?" for _ in data])
        query = f"INSERT INTO {self._table} ({columns}) VALUES ({placeholders})"
        self.cursor.execute(query, tuple(data.values()))
        self.connection.commit()
        return self.cursor.lastrowid

    def update(self, id: int, data: Dict[str, Any]) -> bool:
        set_clause = ", ".join([f"{key} = ?" for key in data])
        query = f"UPDATE {self._table} SET {set_clause} WHERE id = ?"
        self.cursor.execute(query, tuple(data.values()) + (id,))
        self.connection.commit()
        return self.cursor.rowcount > 0

    def delete(self, id: int) -> bool:
        query = f"DELETE FROM {self._table} WHERE id = ?"
        self.cursor.execute(query, (id,))
        self.connection.commit()
        return self.cursor.rowcount > 0

    def create_table(self, table_name: str, columns: Dict[str, str]):
        column_definitions = ", ".join([f"{name} {type}" for name, type in columns.items()])
        query = f"CREATE TABLE IF NOT EXISTS {table_name} (id INTEGER PRIMARY KEY AUTOINCREMENT, {column_definitions})"
        self.cursor.execute(query)
        self.connection.commit()

    def has_many(self, related_table: str, foreign_key: str, local_key: str = "id") -> List[Dict[str, Any]]:
        query = f"SELECT * FROM {related_table} WHERE {foreign_key} = ?"
        self.cursor.execute(query, (self._conditions[0][2],))
        return [dict(row) for row in self.cursor.fetchall()]

    def belongs_to(self, related_table: str, foreign_key: str, local_key: str = "id") -> Optional[Dict[str, Any]]:
        query = f"SELECT * FROM {related_table} WHERE {local_key} = ?"
        self.cursor.execute(query, (foreign_key,))
        row = self.cursor.fetchone()
        return dict(row) if row else None

    def many_to_many(self, join_table: str, related_table: str, related_key: str, local_key: str, foreign_key: str) -> List[Dict[str, Any]]:
        query = f"""
        SELECT DISTINCT {related_table}.* 
        FROM {related_table}
        INNER JOIN {join_table} ON {join_table}.{related_key} = {related_table}.id
        WHERE {join_table}.{foreign_key} = ?
        """
        params = (self._conditions[0][2],) if self._conditions else ()
        self.cursor.execute(query, params)
        return [dict(row) for row in self.cursor.fetchall()]

    def close(self):
        self.connection.close()

# Example usage
if __name__ == "__main__":
    orm = SQLiteORM("example.db")

    # Create tables
    orm.create_table("users", {
        "name": "TEXT",
        "email": "TEXT",
        "age": "INTEGER"
    })

    orm.create_table("projects", {
        "name": "TEXT",
        "description": "TEXT"
    })

    orm.create_table("user_projects", {
        "user_id": "INTEGER",
        "project_id": "INTEGER"
    })

    # Insert users
    user1_id = orm.table("users").insert({"name": "Alice", "email": "alice@example.com", "age": 25})
    user2_id = orm.table("users").insert({"name": "Bob", "email": "bob@example.com", "age": 30})

    # Insert projects
    project1_id = orm.table("projects").insert({"name": "Project A", "description": "First project"})
    project2_id = orm.table("projects").insert({"name": "Project B", "description": "Second project"})

    # Link users and projects (many-to-many relationship)
    orm.table("user_projects").insert({"user_id": user1_id, "project_id": project1_id})
    orm.table("user_projects").insert({"user_id": user1_id, "project_id": project2_id})
    orm.table("user_projects").insert({"user_id": user2_id, "project_id": project1_id})

    # Query projects for a specific user
    user_projects = orm.table("users").where("id", "=", user1_id).many_to_many("user_projects", "projects", "project_id", "user_id", "user_id")
    print("Projects for Alice:", user_projects)

    # Query users for a specific project
    project_users = orm.table("projects").where("id", "=", project1_id).many_to_many("user_projects", "users", "user_id", "project_id", "project_id")
    print("Users in Project A:", project_users)

    orm.close()
